### DISCUSSION TITLE
Is median-scored rule reasonable?

### DISCUSSION THREAD

**Discussion by rezunli96** [yasserfarouk/scml#22] [Q&A] [answered]
I am a little bit skeptical about median-scored rule: since the final score is completely defined by one run, then would the following manipulative behaviors happen? For example if in one simulation an agent think it will going to be very bad, it can abandon this simulation since it knows the scale of this low score wouldn't count. Then it may do something to hurt others. If in another simulation an agent seems to be doing pretty well, then actually it has no incentive to keep being strong and become risk-averse as the scale of this high score will also not be count. I guess in SCML the agents can have some certain estimate about its average-level performances at certain point in a simulation, by checking (1) its current results of contracts executed and (2) other agents' financial report on the bulletin board?

**Comment by yasserfarouk** [comment]
I see your point. Here are my initial thoughts about it:

1. Finding a way to hurt others *using a finite balance and with no possibility of coercion* is an interesting problem in itself and it was the main challenge in the sabotage track of the game in 2019. So I guess an innovation here is useful. Notice that the agent will have some idea about how other agents are doing **in this simulation** but not in the overall competition. This means it is harder for it to target its attacks.  
2. A more serious issue in my opinion, is that the median instead of the mean encourages risk averse behavior. I know for a fact that some agents exhibited such behavior last year by ceasing to trade once they have enough profits accumulated. Nevertheless, the winner agent was actually one that maximized trade and production all the time. May be being too risk averse is not a good idea in std/col after all. For the oneshot track, there is actually no way to stop trading (that is one of the reasons we made the world shallow and forced contract signing). In the tutorial for oneshot, it is clear that even random behavior is better than ceasing trade.
3. The other obvious possibility is to use the mean. We tried that in 2019 and it encourages using unrealistic negotiation agendas with the hope of exploiting bugs in other agents. This was also used by some agents in 2020 but failed because the median is more robust to such manipulations.

This year we considered scoring agents in the average of mean and median this year but it seems like a half hearted measure that would still leave room for such manipulations.

If experience was any guide, winner agents tended to be ones that worked on their negotiation strategy instead of focusing on manipulations of other agent scores.

In summary, we use the median for the following reasons:

1. It seems the lesser of two evils compared with the mean which encourages explorations of bugs in other agents.
2. Hurting others is an interesting problem in itself as long as it is done with limited resources and no coercion.
3. There is no way to stop trading in oneshot. In std, this seems to be less effective as it seems. In collusion, you want to corner the market so manipulations are welcome 😀

This is an important point so let’s discuss it more.

**Reply by rezunli96** [reply]
Thanks for your detailed reply!
1. Yes the agent cannot infer its or others' performances in other simulations. But is it possible that in a simulation an agent at some point comes to know that it had done so poorly that the score will not even close to the overall median, and then just start to hurt others? For example it fails to fulfill many contract, or reach an agreement. Furthermore consider the following agent: for 0.4 probability I adopt a malicious strategy that the only thing I care about is to hurt others. For 0.6 I try to behave strong. Then the median seems to encourage such agent?
2. Yes that is very interesting. Is there any resources/descriptions for past winners? Though there is no way to stop trading in oneshot, it could still encourage some risk-averse behavior in late days right?
3. What do you mean by "exploiting bugs in other agents" and how did that related to discouragement of using mean?

Using the average of mean and median doesn't make sense to me -- I personally advocate using mean.

**Reply by yasserfarouk** [reply]
Thank you for your energy :-)

> But is it possible that in a simulation an agent at some point comes to know that it had done so poorly that the score will not even close to the overall median, and then just start to hurt others?

It is definitely possible but there are some factors that make this less problematic than it seems:
1. By the time you know that you are **behaving badly**, your balance will be low which further reduces your ability to hurt others.
2. By that time, **others also know that you are behaving badly** because financial reports are public and they list your balance, assets, and breaches over time. It is their mistake if you can hurt them given this knowledge given that there can always just turn you down and never trade with you.
3. It is not **easy** to hurt others. Let's think about your two examples: failing to fulfill contracts and failing to reach agreements. The former is not really in your hands as contract execution is forced. You can only fail to fulfill contracts by going bankrupt. This means you need to sign contracts that are beyond your capacity. Given that other agents know your capacity, you should not be able to do that (up to the contracts you already have at this point). The later (failure to reach agreement) is not really a problem for others because there are always other agents that can buy and sell the same product. 

> For 0.6 I try to behave strong. Then the median seems to encourage such agent?

The goal of the competition is to **advance the state of the art in research**. This agent is actually a very good agent for this purpose. It has a good strategy and a malicious one. Both are described and will help others build future strategies. From the point of view of the research community, this agent is doubly good. It would have been bad if the malicious strategy is *trivial* but what I am arguing for is that it is not. Finding a good malicious strategy is not that easy.

> Is there any resources/descriptions for past winners? 

You can read the description of all strategies (with their full source code) on the [scml-agents](https://www.github.com/yasserfarouk/scml-agents) repository. For each agent, you will find a `report.pdf` file in its root folder describing the strategy.

We are planning to have a more academic publication but this will take time.

> Though there is no way to stop trading in oneshot, it could still encourage some risk-averse behavior in late days right?

Yes I agree. The median score encourages being risk-averse near the end in general (and in std/col it does). Nevertheless, to do that you need to first have a good strategy in the beginning then you need to define what does it mean to have **enough profit** in a simulation and what **near the end** means. All of these are open questions. 

>  how did that related to discouragement of using mean?

One common way to maximize your mean is to get very high scores at few simulations. In 2020, we had an agent that got 60% mean profit and 0% median profit (The winner had something like 7% median and mean profits) in std. How did they do that? Simply use crazy negotiation agendas (i.e. when negotiating to sell make the maximum price 100 times the catalog price or 1000). Many agents, normalized their ufuns between zero and one which made them unaware of this very high implied cost and some of them in very few cases accepted these very high prices. This agent could get 1000% profits in relatively few simulations but that was enough to make its mean score extremely high. 


**so the main problem with just the mean is outliers**. It is simply impossible for us to think in advance about all possible ways in which an agent can get an outlier score (very high or very low) in few simulations. For example, in 2019 in which we used the mean, the winning agent could get almost 100 times the score of the second agent because of an exploitation we did not consider in the design allowing them to get extremely high scores in a small number of simulations. The median does not suffer from this problem. **That is why we are considering it the lesser of two evils**.

> I personally advocate using mean

I can see that the mean will encourage taking more risks but it is a less robust estimator in general as it can be affected by outliers. A middle ground solution is to use the mean after applying some appropriate outlier removal method (the simplest is to just chop off lowest and highest 10% of the scores). Using the median is just a special case of this. This approach comes with its own pitfalls. 

**Nevertheless, it may be worth it to consider using the mean after outlier removal without announcing the outlier removal method** This may encourage taking risks without suffering from the outliers issue. 

Finally, we can still control things a little bit by controlling the number of simulations and simulation length. On the extreme of having few very long simulations, the median and the mean will tend to be the same. Participants are not told the number of simulations in advance and we can adjust them accordingly. In general, we will try our best to have long simulations subject to the limitations imposed by the required computation power.

The online competition will be very helpful in understanding how does this evaluation metric affects agent behavior.

**Reply by rezunli96** [reply]
Thank you so much for your exhaustive insights!
It took me some time to think about your points, and I agree on most of them: I also think it is hard to design effective malicious when you had done poorly, and to determine the right time to become risk-averse. 
I am a little bit curious about the 2019 winning agent: what kind of exploitation did it make?
Personally my preference for score rules after our discussion would be: mean with outlier processing > median > vanilla mean

**Reply by yasserfarouk** [reply]
> I am a little bit curious about the 2019 winning agent: what kind of exploitation did it make?

That was caused by the existence of an insurance company which allowed effectively for unlimited injection of funds into the market. This is impossible in the current scml design.

> mean with outlier processing > median > vanilla mean

The problem of `mean with outlier` which made us decide against it is that it will introduce a bias into the evaluation. The reasoning behind that is as follows:

- Market conditions like the distribution of exogenous contracts as well as individual profile conditions like production costs affect the `attainable` profit of each factory.
- To ensure a fair competition, each agent controls each factory the same number of times. As a side effect this makes it possible to use 2-samples tests to ascertain statistical significance of results (but this is just a minor issue).
- If we use outlier removal, then this guarantee will disappear which introduces uncontrolled bias into the evaluation and makes the competition kind of `less fair`. This can happen in either of two ways:
  1. If we remove a complete configuration assignment once a single run of it is deemed an outlier, brittle strategies that work well `on average` but lose too much or cannot make good enough profits in adversarial/opportunistic   circumstances will have an edge.
  2. If we remove only outlier results, the guarantee that `every agent controls every factory the same number of times` is lost.

**Reply by rezunli96** [reply]
> Market conditions like the distribution of exogenous contracts as well as individual profile conditions like production costs affect the attainable profit of each factory.

This can also happen for median right?

>To ensure a fair competition, each agent controls each factory the same number of times. As a side effect this makes it possible to use 2-samples tests to ascertain statistical significance of results (but this is just a minor issue).

You mean the round-robin scheme? What do you mean by "statistical significance"?

>brittle strategies that work well on average but lose too much or cannot make good enough profits in adversarial/opportunistic circumstances will have an edge.

Agreeing with this point. But I guess a debatable question is "what kind of strategies do we consider good (for research purposes)". As in our previous discussion a 0.4 malicious + 0.6 strong agent can be considered a good one. The brittle strategies here may have no reason to be ruled out.

>If we remove only outlier results, the guarantee that every agent controls every factory the same number of times is lost.

So the question is whether this is a reasonable notion of "fairness". First, if it is implemented through round-robin, actually the N! permutations is not explored but only N of them. Not sure whether it will introduce some biases. Second, I do not think it will lost more information than median would do -- for mean with outlier removal it is still aggregating many simulations, but in median it only considers one of them.

**Reply by yasserfarouk** [reply]
Hi, sorry for my late reply but the last few days were crazy 😀.

> This can happen for the median?

Yes, market conditions will still have a big effect but I argue that using the median will make sure that every agent controlled every factory (of the evaluated ones) in every configuration. If we use truncated mean then this is no longer guaranteed. Also truncated mean introduces another ambiguity:

- Truncated mean **per agent** (e.g. remove scores outside some fraction of IQR or SD) *calculated independently for each agent*.
- Truncated **global** mean where we calculate the limits for all the scores not per agent.

Do you have some suggestions about which one seems fairer and would still reduce risk aversion?

> You mean the round-robin scheme? What do you mean by "statistical significance"?

Yes, the round-robin. By the way, our tournament runner can use either round-robin or factorial permutations. In my experience with small numbers of participants (in 2019), there was not much difference between using both of them. Using all possible permutations of agent-factory assignments means that all agents are tested in **almost** the same conditions (notice that even when using factorial, the agent's position itself makes a slight difference). Nevertheless, the bias introduced by round-robin is just that the agents controlling *other* factories are different. I use 2-agents per simulation strategy which means that both of them are equivalent anyway.

By "statistical significance" I meant the, well, statistical significance of the difference between medians reported. When we report that agent A has a median larger than agent B, it tells us how much can we trust this result.

> The brittle strategies here may have no reason to be ruled out.

But it is **not** ruled out. Nothing is really ruled out. As I understand it, the main deciding factors for choosing the evaluation criterion are the following:

1. It should encourage what we consider **good** strategies (playing the game not playing the game-designer). On this dimension, I think we agree that median is better than mean. The remaining question is whether it is better than the global or per-agent truncated mean.
2. It introduces as little bias as possible when comparing agents. It is a competition after all so we need to be sure it is as fair as possible. Moreover, we *know* that market conditions are a big determining factor in this game. This is why I make sure that all agents control all factories. It is also why we use 2-agents per simulation in the tournament. On this dimension, I think that median is better because unlike truncated mean (either version), it guarantees that the *same configurations* are considered for all agents (with all possible assignment).

So it seems to me that in (1) median ~ truncated mean but in (2) median > truncated mean.

> First, if it is implemented through round-robin, actually the N! permutations are not explored but only N of them.

When running the competition, we can control how many factories in the world are *evaluated*. Currently, we use 2 factories per configuration. This means that round robin and all permutations are actually equivalent. This is described in detail in the Tournament Generation appendix in the description doc (Appendix C if I remember correctly).


> I do not think it will be lost more information than median would do -- for mean with outlier removal it is still aggregating many simulations, but in the median, it only considers one of them.

Firstly, the median does not consider a single simulation, because the choice of which simulation is affected by all other simulations.

Nevertheless, we are open to suggestions here. Can you propose a specific way to do the truncated mean that we can evaluate?

These are some possibilities:

- Mean of all simulations per agent after removing the top [and bottom] $\alpha$% per agent
- Mean of all simulations per agent after removing the top [and bottom] $\alpha$% scores globally
- Mean of all simulations per agent after removing scores  $\alpha$ IQR from the median (IQR is calculated per agent)
- Mean of all simulations per agent after removing scores  $\alpha$ IQR from the median (IQR is calculated for all scores globally)
- Mean of $N$ scores around the median per agent (that seems to be better than the above variations). Setting $N$ to $0$ reproduces the median and to something like $25\%$ the number of scores per agent will reproduce the inter-quantile mean.

I do believe that the median is still the most robust estimator and its only problem is low efficiency. I actually went back to last year's scores and checked the value of $N$ (in the last option above) at which the winners (top two agents) change, it turned out to be around $47\%$ of the total number of simulations. This means that using the median or a truncated mean would not have made a difference as long as we removed no less than $6\%$ of the data.

I will open this issue with the game organizers in our next meeting so all input from your side is welcome. We are not in any way married to the median and using a truncated mean is more efficient in data usage I agree (even though slightly less robust).

**Comment by yasserfarouk** [comment, marked as answer]
We will adopt the truncated mean instead of the median in the for I recommended in the last reply. The exact percentages of scores to be removed per agent will not be announced in advance.
